File: gui_view/src/main/java/com/goexp/galgame/gui/view/game/listview/DataViewController.java
package com.goexp.galgame.gui.view.game.listview;

import com.goexp.galgame.common.model.GameState;
import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.task.TaskService;
import com.goexp.galgame.gui.task.game.panel.PanelTask;
import com.goexp.galgame.gui.task.game.panel.node.DefaultItemNode;
import com.goexp.galgame.gui.util.Tags;
import com.goexp.galgame.gui.view.DefaultController;
import com.goexp.galgame.gui.view.game.listview.imglist.ImgListViewController;
import com.goexp.galgame.gui.view.game.listview.sidebar.BrandGroupController;
import com.goexp.galgame.gui.view.game.listview.sidebar.DateGroupController;
import com.goexp.galgame.gui.view.game.listview.sidebar.FilterPanelController;
import com.goexp.galgame.gui.view.game.listview.tableview.TableController;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.collections.transformation.FilteredList;
import javafx.collections.transformation.SortedList;
import javafx.concurrent.Service;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Region;

import java.util.List;
import java.util.function.Predicate;

public class DataViewController extends DefaultController {

    public BooleanProperty reloadProperty = new SimpleBooleanProperty(false);


    /**
     * Controllers
     */
    @FXML
    private ImgListViewController imgViewController;

    @FXML
    public TableController tableViewController;
    @FXML
    private FilterPanelController filterPanelController;
    @FXML
    private BrandGroupController brandGroupController;
    @FXML
    private DateGroupController dateGroupController;


    /**
     * Status bar
     */
    @FXML
    private Label lbItemCount;
    @FXML
    public ProgressBar progessloading;


    /**
     * Toggle
     */
    @FXML
    private ToggleGroup gameViewChange;
    //    @FXML
//    private ToggleButton toggList;
    @FXML
    private ToggleButton toggGrid;
    @FXML
    private ToggleButton toggSmall;
    @FXML
    private ToggleButton toggImg;


    /**
     * main panel
     */
    @FXML
    private TableView<Game> tableView;
    @FXML
    private ListView<Game> smallListSimple;
    //    @FXML
//    private ListView<Game> smallListSimple;
    @FXML
    private Region imgView;


    /**
     * Sidebar
     */
    @FXML
    public Region groupPanel;

    @FXML
    private Region filterPanel;
    @FXML
    private Button btnHide;
    @FXML
    private ListView<DefaultItemNode> cvList;
    public ListView<DefaultItemNode> tagList;


    private FilteredList<Game> filteredGames;

    private Predicate<Game> groupPredicate;

    private Service<List<DefaultItemNode>> groupCVServ = new TaskService<>(() -> new PanelTask.GroupCV(filteredGames));
    private Service<List<DefaultItemNode>> groupTagServ = new TaskService<>(() -> new PanelTask.GroupTag(filteredGames));


    protected void initialize() {
        initSwitchBar();
        initSideBar();
        initCVPanel();
        btnHide.fire();
    }

    private void initCVPanel() {
        cvList.setCellFactory(param -> new ListCell<>() {
            @Override
            protected void updateItem(DefaultItemNode cvItemNode, boolean empty) {
                super.updateItem(cvItemNode, empty);
                setText(null);
                setGraphic(null);

                if (cvItemNode != null && !empty) {
                    setGraphic(new HBox(Tags.toNodes(cvItemNode.title), new Label(String.valueOf(cvItemNode.count))));
                }

            }
        });

        tagList.setCellFactory(param -> new ListCell<>() {
            @Override
            protected void updateItem(DefaultItemNode tagNode, boolean empty) {
                super.updateItem(tagNode, empty);
                setText(null);
                setGraphic(null);

                if (tagNode != null && !empty) {
                    setGraphic(new HBox(Tags.toNodes(tagNode.title), new Label(String.valueOf(tagNode.count))));
                }

            }
        });


        groupCVServ.valueProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue != null) {
                cvList.setItems(FXCollections.observableList(newValue));
            }
        });
        groupTagServ.valueProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue != null) {
                tagList.setItems(FXCollections.observableList(newValue));
            }
        });
    }

    private void initSwitchBar() {
        toggGrid.setUserData(tableView);
        toggSmall.setUserData(smallListSimple);
//        toggList.setUserData(listSimple);
        toggImg.setUserData(imgView);

        gameViewChange.selectedToggleProperty().addListener((observable, oldValue, checkedBox) -> {
            if (checkedBox != null) {
                ((Region) checkedBox.getUserData()).toFront();

            } else {
                gameViewChange.selectToggle(oldValue);
            }

        });
    }

    private void initSideBar() {
        btnHide.setUserData(true);
        btnHide.getProperties().put("groupPanel", groupPanel.getPrefWidth());
        btnHide.setOnAction(event -> {
            var state = (Boolean) btnHide.getUserData();
            var width = (double) btnHide.getProperties().get("groupPanel");

            state = !state;

            groupPanel.setVisible(state);
            groupPanel.setPrefWidth(state ? width : 0);
            btnHide.setUserData(state);

        });


//         #2
//        groupPanel.setExpandedPane(groupPanel.getPanes().get(0));

        initSidebarContentView();
    }

    private void initSidebarContentView() {
        filterPanelController.onSetProperty.addListener((o, old, newV) -> {
            if (newV != null && newV) {
                var load = groupPredicate == null;

                var filterPredicate = filterPanelController.predicate;
                var p = groupPredicate != null ? filterPredicate.and(groupPredicate) : filterPredicate;
                filteredGames.setPredicate(p);

                if(load)
                    setSideBarData(filteredGames);
            }
        });

        brandGroupController.onSetProperty.addListener((observable, oldValue, newValue) -> {
            if (newValue) {
                groupPredicate = brandGroupController.predicate;

                var filterPredicate = filterPanelController.predicate;
                var p = filterPredicate != null ? groupPredicate.and(filterPredicate) : groupPredicate;
                filteredGames.setPredicate(p);
            }
        });
        dateGroupController.onSetProperty.addListener((observable, oldValue, newValue) -> {
            if (newValue) {
                groupPredicate = dateGroupController.predicate;

                var filterPredicate = filterPanelController.predicate;
                var p = filterPredicate != null ? groupPredicate.and(filterPredicate) : groupPredicate;
                filteredGames.setPredicate(p);
            }
        });
    }


    public void load(ObservableList<Game> games, Predicate<Game> initPredicate) {

        filterPanel.setVisible(true);

        groupPredicate = null;

        filteredGames = new FilteredList<>(games);
        filteredGames.predicateProperty().addListener((o, old, newV) -> {
            if (newV != null) {
                tableView.scrollTo(0);
                smallListSimple.scrollTo(0);
                resetCount(filteredGames);
            }
        });

        // set filter
        filteredGames.setPredicate(initPredicate);

        // set defaultPredicate
        filterPanelController.predicate = initPredicate;

        var sortedData = new SortedList<>(filteredGames);
        sortedData.comparatorProperty().bind(tableView.comparatorProperty());

        loadItems(sortedData);

        setSideBarData(filteredGames);
    }


    public void load(ObservableList<Game> games) {
        Predicate<Game> defaultP = g -> g.state.get().value > GameState.BLOCK.value && !(g.star > 0 && g.star < 3);
        load(games, defaultP);
    }

    private void setSideBarData(FilteredList<Game> filteredGames) {
        dateGroupController.init(filteredGames);
        brandGroupController.init(filteredGames);

        groupCVServ.restart();
        groupTagServ.restart();
    }

    private void loadItems(SortedList<Game> sortedData) {

        tableView.setItems(sortedData);
        tableView.scrollTo(0);

        smallListSimple.setItems(sortedData);
        smallListSimple.scrollTo(0);

        imgViewController.load(sortedData);
    }

//    private void loadItems(ObservableList<Game> sortedData) {
//        resetCount(sortedData);
//        tableView.setItems(sortedData);
//        tableView.scrollTo(0);
//
//        listSimple.setItems(sortedData);
//        listSimple.scrollTo(0);
//    }

    private void resetCount(List<Game> filteredGames) {
        lbItemCount.setText(String.format("%d 件", filteredGames.size()));
    }


    @FXML
    private void reload_OnAction(ActionEvent event) {

        reloadProperty.set(true);
        reloadProperty.set(false);

    }
}

File: gui_view/src/main/scala/com/goexp/galgame/gui/view/game/listview/DataViewController.scala
package com.goexp.galgame.gui.view.game.listview

import java.util
import java.util.function.Predicate

import com.goexp.galgame.common.model.game.GameState
import com.goexp.galgame.gui.model.Game
import com.goexp.galgame.gui.task.TaskService
import com.goexp.galgame.gui.task.game.panel.group.node.DefaultItem
import com.goexp.galgame.gui.task.game.panel.group.{ByCV, ByTag}
import com.goexp.galgame.gui.util.Tags
import com.goexp.galgame.gui.view.DefaultController
import com.goexp.galgame.gui.view.game.listview.imglist.ImgListViewController
import com.goexp.galgame.gui.view.game.listview.sidebar.{BrandGroupController, DateGroupController, FilterPanelController}
import com.goexp.galgame.gui.view.game.listview.tableview.TableController
import javafx.beans.property.SimpleBooleanProperty
import javafx.collections.transformation.{FilteredList, SortedList}
import javafx.collections.{FXCollections, ObservableList}
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.control._
import javafx.scene.layout.{HBox, Region}

class DataViewController extends DefaultController {
  final val reloadProperty = new SimpleBooleanProperty(false)
  /**
    * Controllers
    */
  @FXML private var imgViewController: ImgListViewController = _
  @FXML var tableViewController: TableController = _
  @FXML private var filterPanelController: FilterPanelController = _
  @FXML private var brandGroupController: BrandGroupController = _
  @FXML private var dateGroupController: DateGroupController = _
  /**
    * Status bar
    */
  @FXML private var lbItemCount: Label = _
  @FXML var progessloading: ProgressBar = _
  /**
    * Toggle
    */
  @FXML private var gameViewChange: ToggleGroup = _
  //    @FXML
  //    private ToggleButton toggList;
  @FXML private var toggGrid: ToggleButton = _
  @FXML private var toggSmall: ToggleButton = _
  @FXML private var toggImg: ToggleButton = _
  /**
    * main panel
    */
  @FXML private var tableView: TableView[Game] = _
  @FXML private var smallListSimple: ListView[Game] = _
  @FXML private var imgView: Region = _
  /**
    * Sidebar
    */
  @FXML private var groupPanel: Region = _
  @FXML private var filterPanel: Region = _
  @FXML private var btnHide: Button = _
  @FXML private var cvList: ListView[DefaultItem] = _
  @FXML private var tagList: ListView[DefaultItem] = _
  private var filteredGames: FilteredList[Game] = _
  private var groupPredicate: Predicate[Game] = _

  final private val groupCVServ = TaskService(() => new ByCV(filteredGames))
  final private val groupTagServ = TaskService(() => new ByTag(filteredGames))

  override protected def initialize() = {
    initSwitchBar()
    initSideBar()
    initGroupPanel()
    btnHide.fire()
  }

  private def initGroupPanel() = {
    cvList.setCellFactory(_ => new ListCell[DefaultItem]() {
      override protected def updateItem(cvItemNode: DefaultItem, empty: Boolean) = {
        super.updateItem(cvItemNode, empty)
        setText(null)
        setGraphic(null)
        if (cvItemNode != null && !empty)
          setGraphic(new HBox(Tags.toNodes(cvItemNode.title), new Label(String.valueOf(cvItemNode.count))))
      }
    })
    tagList.setCellFactory(_ => new ListCell[DefaultItem]() {
      override protected def updateItem(tagNode: DefaultItem, empty: Boolean) = {
        super.updateItem(tagNode, empty)
        setText(null)
        setGraphic(null)
        if (tagNode != null && !empty) setGraphic(new HBox(Tags.toNodes(tagNode.title), new Label(String.valueOf(tagNode.count))))
      }
    })
    groupCVServ.valueProperty.addListener((_, _, newValue) => {
      if (newValue != null) cvList.setItems(FXCollections.observableList(newValue))
    })
    groupTagServ.valueProperty.addListener((_, _, newValue) => {
      if (newValue != null) tagList.setItems(FXCollections.observableList(newValue))
    })
  }

  private def initSwitchBar() = {
    toggGrid.setUserData(tableView)
    toggSmall.setUserData(smallListSimple)
    toggImg.setUserData(imgView)
    gameViewChange.selectedToggleProperty.addListener((_, oldValue, checkedBox) => {
      if (checkedBox != null) checkedBox.getUserData.asInstanceOf[Region].toFront()
      else gameViewChange.selectToggle(oldValue)
    })
  }

  private def initSideBar() = {
    btnHide.setUserData(true)
    btnHide.getProperties.put("groupPanel", groupPanel.getPrefWidth)
    btnHide.setOnAction(_ => {
      var state = btnHide.getUserData.asInstanceOf[Boolean]
      val width = btnHide.getProperties.get("groupPanel").asInstanceOf[Double]
      state = !state
      groupPanel.setVisible(state)
      groupPanel.setPrefWidth(if (state) width else 0)
      btnHide.setUserData(state)
    })
    initSidebarContentView()
  }

  private def initSidebarContentView() = {
    filterPanelController.onSetProperty.addListener((_, _, newV) => {
      if (newV) {
        val load = groupPredicate == null
        val filterPredicate = filterPanelController.predicate
        val p = if (groupPredicate != null) filterPredicate.and(groupPredicate)
        else filterPredicate
        filteredGames.setPredicate(p)
        recount()
        if (load) setSideBarData(filteredGames)
      }

    })
    brandGroupController.onSetProperty.addListener((_, _, newValue) => {
      if (newValue) {
        groupPredicate = brandGroupController.predicate
        val filterPredicate = filterPanelController.predicate
        val p = if (filterPredicate != null) groupPredicate.and(filterPredicate)
        else groupPredicate
        filteredGames.setPredicate(p)
        recount()
      }

    })
    dateGroupController.onSetProperty.addListener((_, _, newValue) => {
      if (newValue) {
        groupPredicate = dateGroupController.predicate
        val filterPredicate = filterPanelController.predicate
        val p = if (filterPredicate != null) groupPredicate.and(filterPredicate)
        else groupPredicate
        filteredGames.setPredicate(p)
        recount()
      }

    })
  }

  def recount(): Unit = {
    tableView.scrollTo(0)
    smallListSimple.scrollTo(0)
    resetCount(filteredGames)
  }

  def load(games: ObservableList[Game], initPredicate: Predicate[Game]): Unit = {
    filterPanel.setVisible(true)
    groupPredicate = null
    filteredGames = new FilteredList(games)

    // set filter
    filteredGames.setPredicate(initPredicate)
    recount()

    // set defaultPredicate
    filterPanelController.predicate = initPredicate
    val sortedData = new SortedList[Game](filteredGames)
    sortedData.comparatorProperty.bind(tableView.comparatorProperty)
    loadItems(sortedData)
    setSideBarData(filteredGames)
  }

  def load(games: ObservableList[Game]): Unit = {
    val defaultP: Predicate[Game] = (g: Game) => (g.state.get ne GameState.SAME) &&
      (g.state.get ne GameState.BLOCK) &&
      !(g.star > 0 &&
        g.star < 3)

    load(games, defaultP)
  }

  private def setSideBarData(filteredGames: FilteredList[Game]) = {
    dateGroupController.init(filteredGames)
    brandGroupController.init(filteredGames)
    groupCVServ.restart()
    groupTagServ.restart()
  }

  private def loadItems(sortedData: SortedList[Game]) = {
    tableView.setItems(sortedData)
    tableView.scrollTo(0)
    smallListSimple.setItems(sortedData)
    smallListSimple.scrollTo(0)
  }

  private def resetCount(filteredGames: util.List[Game]) =
    lbItemCount.setText(s"${filteredGames.size} 件")

  @FXML private def reload_OnAction(event: ActionEvent) = {
    reloadProperty.set(true)
    reloadProperty.set(false)
  }
}